pyomo/contrib/doe/DoE_package.py

```diff
@@ -0,0 +1,1812 @@
+        if (self.prior_FIM is not None):
+            try:
+                fim = fim + self.prior_FIM
+                print('Existed information has been added.')
+            except:
+                raise ValueError('Check the shape of prior FIM')
+
+        if np.linalg.cond(fim) > self.max_condition_number:
+            print("Warning: FIM is near singular.")
+
+        # call private methods
+        self.__print_FIM_info(fim, dv_set=dv_values)
+        if self.result is not None:
+            self.__solver_info()
+
+        # if given store file name, store the FIM
+        if (self.store_FIM is not None):
+            self.__store_FIM()
+
+    def extract_FIM(self, m, dv_set, result, obj=None, add_fim=False):
+        '''
+        Extract FIM from an invasive model
+
+        Args:
+            m: model
+            dv_set: design variable value dictionary
+            result: problem solver status by IPOPT
+            obj: chosen from 'det' and 'trace'
+            add_fim: if the given FIM needs to be added. Do not add for optimize_doe().
+
+        Return:
+            fim_info: a FIM dictionary
+                ~['FIM']: FIM itself
+                ~[design variable name]: design variable values at each time point
+                ~['Trace']: Trace
+                ~['Determinant']: determinant
+                ~['Condition number:']: condition number
+                ~['Minimal eigen value:']: minimal eigen value
+                ~['Eigen values:']: all eigen values
+                ~['Eigen vectors:']: all eigen vectors
+            model_info: model solutions dictionary
+                ~['obj']: objective function value
+                ~['det']: determinant calculated by the model (different from FIM_info['det'] which
+                is calculated by numpy)
+                -['trace']: trace calculated by the model
+                -[design variable name]: design variable solution
+            solver_status: a solver infomation dictionary
+                ~['square']: square result solver status
+                -['doe']: doe result solver status
+        '''
+        self.result = result
+        self.obj = obj
+        no_para = len(self.para_name)
+
+        # Extract FIM infomation
+        FIM = np.ones((no_para, no_para))
+        for n1, name1 in enumerate(self.para_name):
+            for n2, name2 in enumerate(self.para_name):
+                FIM[n1, n2] = value(m.FIM[name1, name2]) / self.fim_scale_constant_value
+
+        # add prior information
+        if add_fim:
+            if (self.prior_FIM is not None):
+                try:
+                    FIM = FIM + self.prior_FIM
+                    print('Existed information has been added.')
+                except:
+                    raise ValueError('Check the shape of prior FIM')
+
+        # call private methods
+        self.__print_FIM_info(FIM, dv_set=dv_set)
+        self.__solution_info(m, dv_set)
+        self.__solver_info()
```

Comment:
Another method did "if self.result is not None:". Is this needed here?

Reply:
Yes, it's because this is only valid for simultaneous ways. Sequential ways and IPOPT run several models for one experiment, so in Grid search(), there will be thousands of IPOPT results. I don't check the convergence of them for now.